docs/contributing/design-documents.rst

```diff
@@ -0,0 +1,74 @@
+:title: Design Documents
+:description: Information necessary for a feature design document.
+
+.. _design-documents:
+
+Design Documents
+================
+
+Before submitting a pull request which is expected to significantly change the
+behavior of Deis (such as when contributing a new feature or refactor),
+contributors should first open an issue representing a design document.
+
+Goals
+-----
+
+Design documents were adopted in order to:
+
+* Involve stakeholders as early as possible in a feature's development
+* Ensure code changes accomplish the original motivations and design goals
+* Establish clear acceptance criteria for a feature or change
+* Enforce test-driven design methodology and automated test coverage
+
+Contents
+--------
+
+Design document issues should be named ``Design Doc: <change description>`` and
+contain the following sections.
+
+Goal
+^^^^
+
+This section should briefly describe the proposed change and the motivations
+behind the change. Tests will be written to ensure this design goal is met by
+the change.
+
+This section will typically reference a separate GitHub issue tracking
+the feature or change, which will typically be assigned to a release milestone.
+
+Code Changes
+^^^^^^^^^^^^
+
+This section should detail the code changes necessary to accomplish the change,
```

Comment:
Commit bodies are good, but this is meant to be more of a discussion, forcing people to think through the change - "we need to change this code because of how it's used here, meaning it should be refactored as an interface".